### PR TITLE
Configure health check timeouts

### PR DESCRIPTION
**Pull request by willmostly**
## Description
Add timeout parameters for INFO_API and JDBC health checks.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x ) Release notes are required, with the following suggested text:
Add timeout parameters for INFO_API and JDBC health checks.

### CODE REVIEW

File: docs/installation.md [hunk 1]
```diff
@@ -244,3 +244,51 @@ Kubernetes updates the mounted file when the ConfigMap is updated.
 Standard Helm oœptions such as `replicaCount`, `image`, `imagePullSecrets`, 
 `service`, `ingress` and `resources` are supported. These are defined in 
 `helm/values.yaml`. 
+
+### Health Checks
+
+Trino Gateway checks the health of each backend and deactivates it if unhealthy.
```
Comment (andythsu) [inline docs/installation.md:250]: I'd probably bold "deactivates it if unhealthy" because it's an important behavior and sometimes confuses people why their backends are deactivated.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java [hunk 2]
```diff
@@ -47,18 +63,42 @@ public ClusterStats monitor(ProxyBackendConfiguration backend)
...
                 .build();
-
         try {
             ServerInfo serverInfo = client.execute(request, SERVER_INFO_JSON_RESPONSE_HANDLER);
             return !serverInfo.isStarting();
         }
+        catch (UnexpectedResponseException e) {
+            if (shouldRetry(e.getStatusCode())) {
+                if (retriesRemaining > 0) {
+                    return isReadyStatus(baseUrl, retriesRemaining - 1);
```
Comment (andythsu) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java:82]: I'd log something to inform that the retries are happening. Maybe something like `log.warn("received an retriable error, retrying...")

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java [hunk 3]
```diff
@@ -47,18 +63,42 @@ public ClusterStats monitor(ProxyBackendConfiguration backend)
...
+                }
+            }
+            log.warn(e, "Health check failed with non-retryable response. %s", e.toString());
```
Comment (andythsu) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java:85]: Not only does non-retryable response get here. We will also end up in this statement if `retries == 0`. We should add something to distinguish between these two states

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsJdbcMonitor.java [hunk 4]
```diff
@@ -73,8 +81,16 @@ public ClusterStats monitor(ProxyBackendConfiguration backend)
...
+            networkTimeoutMillis.ifPresent(t -> {
+                try {
+                    conn.setNetworkTimeout(Executors.newSingleThreadExecutor(), 10); //executor is unused in TrinoConnection.setNetworkTimeout
+                }
+                catch (SQLException e) {
+                    log.error(e, "Could not set network timeout for JDBC health monitor");
+                }
+            });
+            PreparedStatement stmt = SimpleTimeLimiter.create(Executors.newSingleThreadExecutor()).callWithTimeout(
+                    () -> conn.prepareStatement(STATE_QUERY), max(10, 1000 * networkTimeoutMillis.orElse(0)) + 1, TimeUnit.SECONDS);
```
Comment (andythsu) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsJdbcMonitor.java:93]: we should add a comment here to explain the 10 seconds magic number here. Basically this is querying the `system.runtime` table and we don't expect more than 10 seconds, and hence it is "hardcoded" here and not configurable. There were some asks to make it configurable, which is something we can do, but it doesn't add much value since 10 seconds should be more than enough

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java (hunk 3 above)
Comment (vishalya) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java:85]: Why is this a warning and not an error? even though it's non retry-able response.

**Comment by mosabua** [comment]
Let us know when this is rebased and ready again @willmostly

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsJdbcMonitor.java (hunk 4 above)
Comment (willmostly) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsJdbcMonitor.java:93]: This is making it configurable - I was using 10 as the default, but did so in a confusing way, and screwed up the conversion from millis to seconds to boot.

Comment (willmostly) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsJdbcMonitor.java]: this was reverted since the network config was superseded by https://github.com/trinodb/trino-gateway/pull/436

**Comment by willmostly** [comment]
Note that the configuration options introduced here are superseded by https://github.com/trinodb/trino-gateway/pull/436. Retries are now the only thing being added

**Comment by willmostly** [comment]
@andythsu and @vishalya this is ready for another look

**Review by andythsu — APPROVED** [review]
I'm good with this one! 🚢 it

**Review by mosabua — COMMENTED** [review]
Please add empty lines after titles, before and after codeblocks and lists and verify that the docs actually render properly

**Review by mosabua — APPROVED** [review]

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java [hunk 5]
```diff
@@ -50,18 +56,48 @@ public ClusterStats monitor(ProxyBackendConfiguration backend)
     }
 
     private boolean isReadyStatus(String baseUrl)
+    {
+        return isReadyStatus(baseUrl, retries);
+    }
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java:61]: This method seems redundant. L52 could call `isReadyStatus(String baseUrl, int retriesRemaining)` instead.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java [hunk 6]
```diff
@@ -50,18 +56,48 @@ public ClusterStats monitor(ProxyBackendConfiguration backend)
...
         Request request = prepareGet()
                 .setUri(uriBuilderFrom(URI.create(baseUrl)).appendPath("/v1/info").build())
                 .build();
-
         try {
             ServerInfo serverInfo = client.execute(request, SERVER_INFO_JSON_RESPONSE_HANDLER);
             return !serverInfo.isStarting();
         }
+        catch (UnexpectedResponseException e) {
+            if (shouldRetry(e.getStatusCode())) {
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java:73]: Please consider replacing this retry logic with [Failsafe](https://failsafe.dev/). We widely use this library in Trino.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java [hunk 7]
```diff
@@ -50,18 +56,48 @@ public ClusterStats monitor(ProxyBackendConfiguration backend)
...
+    public static boolean shouldRetry(int statusCode)
+    {
+        switch (statusCode) {
+            case HTTP_BAD_GATEWAY:
+            case HTTP_UNAVAILABLE:
+            case HTTP_GATEWAY_TIMEOUT:
+                return true;
+            default:
+                return false;
+        }
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java:101]: We could use switch expression.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java [hunk 8]
```diff
@@ -16,28 +16,34 @@
...
 import static java.util.Objects.requireNonNull;
 
 public class ClusterStatsInfoApiMonitor
         implements ClusterStatsMonitor
 {
-    private static final Logger log = LoggerFactory.getLogger(ClusterStatsInfoApiMonitor.class);
+    private static final Logger log = Logger.get(ClusterStatsInfoApiMonitor.class);
     private static final JsonResponseHandler<ServerInfo> SERVER_INFO_JSON_RESPONSE_HANDLER = createJsonResponseHandler(jsonCodec(ServerInfo.class));
     private final HttpClient client;
+    int retries;
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java:41]: Add `private final`.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsJdbcMonitor.java [hunk 9]
```diff
@@ -43,7 +44,7 @@ public class ClusterStatsJdbcMonitor
             + "WHERE user != ? AND date_diff('hour',created,now()) <= 1 "
             + "GROUP BY state";
 
-    public ClusterStatsJdbcMonitor(BackendStateConfiguration backendStateConfiguration)
+    public ClusterStatsJdbcMonitor(BackendStateConfiguration backendStateConfiguration, MonitorConfiguration monitorConfiguration)
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsJdbcMonitor.java:47]: `MonitorConfiguration monitorConfiguration` is unused.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsJdbcMonitor.java [hunk 10]
```diff
@@ -73,8 +74,8 @@ public ClusterStats monitor(ProxyBackendConfiguration backend)
         }
 
         try (Connection conn = DriverManager.getConnection(jdbcUrl, properties)) {
-            PreparedStatement stmt = SimpleTimeLimiter.create(Executors.newSingleThreadExecutor())
-                    .callWithTimeout(() -> conn.prepareStatement(STATE_QUERY), 10, TimeUnit.SECONDS);
+            PreparedStatement stmt = SimpleTimeLimiter.create(Executors.newSingleThreadExecutor()).callWithTimeout(
+                    () -> conn.prepareStatement(STATE_QUERY), 10, TimeUnit.SECONDS);
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsJdbcMonitor.java:78]: This change seems unrelated to "Configure health check timeouts".

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java [hunk 11]
```diff
@@ -50,18 +56,48 @@ public ClusterStats monitor(ProxyBackendConfiguration backend)
...
                 .build();
-
         try {
             ServerInfo serverInfo = client.execute(request, SERVER_INFO_JSON_RESPONSE_HANDLER);
             return !serverInfo.isStarting();
         }
+        catch (UnexpectedResponseException e) {
+            if (shouldRetry(e.getStatusCode())) {
+                if (retriesRemaining > 0) {
+                    log.warn("Retrying health check on error: %s, ", e.toString());
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsInfoApiMonitor.java:75]: `,` after `%s` seems redundant.
